feat: AWS Lambda authorizer support

diegocstn opened the pull request:
*Description of changes:*
This PR introduces the changes to support for AWS Lambda authorizer.
Customers will be able to use the new `AWS_LAMBDA` authorization type by providing an instance of a type conforming to the `AWSLambdaAuthProvider` protocol, similarly to as they currently do for OIDC.

Example of dx:
```swift

struct MyAuthProvider: AWSLambdaAuthProvider {
    func getLatestAuthToken() -> String {
        "custom-lambda-token"
    }
}
``

let authProvider = MyAuthProvider()
let serviceConfig = try AWSAppSyncServiceConfig()
let config = try AWSAppSyncClientConfiguration(
    appSyncServiceConfig: serviceConfig,
    awsLambdaAuthProvider: authProvider)

let appSyncClient = try AWSAppSyncClient(appSyncConfig: config)
```

**References**
Amplify iOS PR: https://github.com/aws-amplify/amplify-ios/pull/1334
Amplify Android PR:  https://github.com/aws-amplify/amplify-android/pull/1412
AppSync SDK Android PR:  https://github.com/awslabs/aws-mobile-appsync-sdk-android/pull/358

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

royjit reviewed (approved):
LGTM, only concern I had is regarding inheriting the protocol AWSLambdaAuthProvider from `AWSOIDCAuthProvider`. Please provide comments on why it was done so

diegocstn commented:
@royjit The idea was to re-use the OIDC interceptor given that they both just add a token as the authorization header value. Changed to use a custom interceptor as I did for Amplify and removed the protocol conformance. 
I'm gonna create a story to refactor the auth interceptors across packages. I believe we can re-use some code there.
